nanofi/src/sitetosite/CRawSocketProtocol.c

```diff
@@ -0,0 +1,1026 @@
+    RespondCode code;
+    if(readResponse(client, &code) <= 0) {
+      return -1;
+    }
+
+    RespondCodeContext *resCode = getRespondCodeContext(code);
+
+    if (resCode->hasDescription) {
+      uint32_t utflen;
+      int ret = readUTFLen(&utflen, client->_peer->_stream);
+      if (ret <= 0)
+        return -1;
+      memset(client->_description_buffer, 0, utflen+1);
+      ret = readUTF(client->_description_buffer, utflen, client->_peer->_stream);
+      if (ret <= 0)
+        return -1;
+    }
+
+    // we've sent a FINISH_TRANSACTION. Now we'll wait for the peer to send a 'Confirm Transaction' response
+    if (code == CONFIRM_TRANSACTION) {
+      logc(debug, "Site2Site transaction %s peer confirm transaction with CRC %s", transactionID, client->_description_buffer);
+
+      if (client->_currentVersion > 3) {
+        int64_t crcValue = getCRC(transaction);
+        char crc[40];
+        memset(crc, 0, 40);
+        sprintf(crc, "%lld", crcValue);
+
+        if (strcmp(client->_description_buffer, crc) == 0) {
+          logc(debug, "Site2Site transaction %s CRC matched", transactionID);
+          if(writeResponse(client, CONFIRM_TRANSACTION, "CONFIRM_TRANSACTION") <= 0) {
+            return -1;
+          }
+          transaction->_state = TRANSACTION_CONFIRMED;
+          return 0;
+        } else {
+          logc(warn, "Site2Site transaction %s CRC not matched %s", transactionID, crc);
+          writeResponse(client, BAD_CHECKSUM, "BAD_CHECKSUM");
+          return -1;
+        }
+      }
+      if (writeResponse(client, CONFIRM_TRANSACTION, "CONFIRM_TRANSACTION") <= 0) {
+        return -1;
+      }
+      transaction->_state = TRANSACTION_CONFIRMED;
+      return 0;
+    } else {
+      logc(debug, "Site2Site transaction %s peer unknown respond code %d", transactionID, code);
+      return -1;
+    }
+  }
+}
+
+  int16_t sendPacket(struct CRawSiteToSiteClient * client, const char * transactionID, CDataPacket *packet, flow_file_record * ff) {
+
+    if (client->_peer_state != READY) {
+      bootstrap(client);
+    }
+
+    if (client->_peer_state != READY) {
+      return -1;
+    }
+    CTransaction* transaction = findTransaction(client, transactionID);
+
+    if (!transaction) {
+      return -1;
+    }
+
+    if (getState(transaction) != TRANSACTION_STARTED && getState(transaction) != DATA_EXCHANGED) {
+      logc(warn, "Site2Site transaction %s is not at started or exchanged state", transactionID);
+      return -1;
+    }
+
+    if (getDirection(transaction) != SEND) {
+      logc(warn, "Site2Site transaction %s direction is wrong", transactionID);
+      return -1;
+    }
+
+    int ret;
+
+    if (transaction->current_transfers_ > 0) {
+      ret = writeResponse(client, CONTINUE_TRANSACTION, "CONTINUE_TRANSACTION");
+      if (ret <= 0) {
+        return -1;
+      }
+    }
+    // start to read the packet
+    uint32_t numAttributes = packet->_attributes->size;
+    ret = write_uint32t(transaction, numAttributes);
```

Comment:
Is write_uint32t a variation of write_uint32_t?

Reply:
Short answer: yes.
Long answer: I needed dedicated write function for transaction to maintain crc checksum.
However there is no overload in C, so the function to write to transaction has to have a different name than the one that's used to write to streams.